### ISSUE TITLE
Implement IPv6 support

### ISSUE THREAD

**Issue by erkinalp** [Genymobile/gnirehtet#285]
gnirehteT currently only supports IPv4. IPv6 is becoming increasingly commoner.

**Comment by rom1v** [comment]
Yes, but that's a huge work, and I have absolutely no time to do that. :/

**Comment by yezhengli-Mr9** [comment]
TO DO
* [src/main/java/com/genymobile/gnirehtet/relay/](https://github.com/Genymobile/gnirehtet/tree/master/relay-java/src/main/java/com/genymobile/gnirehtet/relay)`IPv6Packet.java`
*

**Comment by yezhengli-Mr9** [comment]
# TO DO for rust part
`brew install gnirehtet` is rust version according to [README.md](https://github.com/Genymobile/gnirehtet).
*  new files
   + in [elay-rust/src/relay/](https://github.com/Genymobile/gnirehtet/tree/master/relay-rust/src/relay)`ipv6_packet.rs` add `pub struct Ipv6Packet<'a> {`
   
   + in [elay-rust/src/relay/](https://github.com/Genymobile/gnirehtet/tree/master/relay-rust/src/relay)`ipv6_packet_buffer.rs` add `pub struct Ipv6PacketBuffer {`
* Consequentially, add
  + in [packetizer.rs](https://github.com/Genymobile/gnirehtet/tree/master/relay-rust/src/relay/packetizer.rs)
      + `pub fn packetize_empty_payload(&mut self) -> Ipv6Packet {`
      + `pub fn packetize<R: DatagramReceiver>(&mut self, source: &mut R) -> io::Result<Ipv6Packet> {`
      + ` pub fn ipv6_header_mut(&mut self) -> Ipv6HeaderMut {`
      + `pub fn inflate(&mut self, packet_length: u16) -> Ipv6Packet {`
  + in [udp_connection.java](https://github.com/Genymobile/gnirehtet/tree/master/relay-rust/src/relay/udp_connection.rs)
      + `pub fn create(selector: &mut Selector, id: ConnectionId, client: Weak<RefCell<Client>>, ipv6_header: Ipv6 Header, transport_header: TransportHeader,)`
      + `fn send_to_network(&mut self, selector: &mut Selector, _: &mut ClientChannel, ipv6_packet: &ipv6Packet,)`, 
  + in [tcp_connection.rs](https://github.com/Genymobile/gnirehtet/tree/master/relay-rust/src/relay/tcp_connection.rs)
      + `pub fn create(selector: &mut Selector, id: ConnectionId, client: Weak<RefCell<Client>>, ipv6_header: Ipv6Header, transport_header: TransportHeader,) -> io::Result<Rc<RefCell<Self>>> {`, 
      + `fn send_to_client(client: &Weak<RefCell<Client>>, selector: &mut Selector, ipv6_packet: &Ipv6Packet, ) -> io::Result<()> {`, 
      + `fn tcp_header_of_packet<'a>(ipv6_packet: &'a Ipv6Packet) -> TcpHeader<'a> {` , 
      + `fn handle_packet(&mut self, selector: &mut Selector, client_channel: &mut ClientChannel, ipv6_packet: &Ipv6Packet,) {`, 
      + `fn handle_first_packet(&mut self, selector: &mut Selector, client_channel: &mut ClientChannel, ipv6_packet: &Ipv6Packet,) {`, 
      + `fn handle_duplicate_syn(&mut self, selector: &mut Selector, client_channel: &mut ClientChannel, ipv6_packet: &Ipv6Packet,) {`, 
      + `fn handle_ack(&mut self, _selector: &mut Selector, _client_channel: &mut ClientChannel, ipv6_packet: &Ipv6Packet,) {`, 
      + `fn send_to_network(&mut self, selector: &mut Selector, client_channel: &mut ClientChannel, ipv6_packet: &Ipv6Packet,) {` 
  + in [connection.rs](https://github.com/Genymobile/gnirehtet/tree/master/relay-rust/src/relay/connection.rs) and [router.rs](https://github.com/Genymobile/gnirehtet/tree/master/relay-rust/src/relay/router.rs) `pub fn send_to_network(&mut self, selector: &mut Selector, client_channel: &mut ClientChannel, ipv6_packet: &Ipv6Packet,) {`

**Comment by yezhengli-Mr9** [comment]
> gnirehteT currently only supports IPv4. IPv6 is becoming increasingly commoner.

Hi @rom1v , any reference code for [IPv4Packet.java](https://github.com/Genymobile/gnirehtet/blob/master/relay-java/src/main/java/com/genymobile/gnirehtet/relay/IPv4Packet.java)? Maybe original reference has `IPv6Packet` as well.

**Comment by rom1v** [comment]
> Any reference code for [IPv4Packet.java](https://github.com/Genymobile/gnirehtet/blob/master/relay-java/src/main/java/com/genymobile/gnirehtet/relay/IPv4Packet.java)?

This is the original code. Also, adding IPv6 support requires more than just some renaming :wink:

**Comment by yezhengli-Mr9** [comment]
> # TO DO for java part
## For java

1. [TODOs for IPv6 headers and Tests as well as new classes](https://github.com/Genymobile/gnirehtet/commit/dfa5615753a04b3de0c3b3f44dac1cf52d7ac5ea). Pay attention to IPv6 format.

2. [Packetizer and their tests](https://github.com/Genymobile/gnirehtet/commit/b03230ba233416de61093805c293aa85d0a933bb) (messed up one file with 1, one file with 3). 

3. [Connection (TCP, UDP) and their tests](https://github.com/Genymobile/gnirehtet/commit/eb6c6fb5191f37bde4b53bda0429ca0da9fe567f). UDP (one week) seems easier than TCP (two weeks). 

4. More tests.

**Comment by stepnoy** [comment]
Is there any chance to merge it? Or ipv6 won't work anyway?
